projects/valtimo/form-view-model/src/lib/components/form-view-model/form-view-model.component.ts

```diff
@@ -0,0 +1,247 @@
+        this.viewModelService
+          .submitViewModel(formName, taskInstanceId, submission.data)
+          .pipe(take(1))
+          .subscribe({
+            next: response => {
+              callback(null, submission);
+            },
+            error: error => {
+              this.handleFormError(error);
+              callback({message: error.error.error, component: null}, null);
+            },
+          });
+      });
+  }
+
+  private handleFormError(error: HttpErrorResponse) {
+    const formInstance = this.formio.formio;
+    const component = formInstance.getComponent(error.error?.component);
+    if (component == null) {
+      this.errors$.next([error.error.error]);
+    } else {
+      setTimeout(() => {
+        component?.setCustomValidity(error.error.error);
+      }, 500);
+    }
+  }
+
+  public onSubmit(submission: FormioSubmission): void {
+    this.formSubmit.next(submission);
+  }
+
+  public onChange(object: any): void {
+    if (object.data) {
+      this.change$.next(object);
+    }
+
+    if (object.changed) {
+      this.submission$.next(this.submission);
+      this.handleChanges();
+    }
+  }
+
+  public loadInitialViewModel() {
+    combineLatest([this.formName$, this.taskInstanceId$])
+      .pipe(take(1))
+      .subscribe(([formName, taskInstanceId]) => {
+        this.viewModelService.getViewModel(formName, taskInstanceId).subscribe(viewModel => {
+          this.change$.pipe(take(1)).subscribe(change => {
+            this.loading$.next(false);
+          });
+          this.submission$.next({data: viewModel});
+        });
+      });
+  }
+
+  public updateViewModel() {
+    this.loading$.pipe(take(1)).subscribe(updating => {
+      if (!updating) {
+        this.loading$.next(true);
+        combineLatest([this.formName$, this.taskInstanceId$, this.change$])
+          .pipe(take(1))
+          .subscribe(([formName, taskInstanceId, change]) => {
+            this.viewModelService.updateViewModel(formName, taskInstanceId, change.data).subscribe({
+              next: viewModel => {
+                this.submission$.next({data: viewModel});
+                this.change$.pipe(take(1)).subscribe(change => {
+                  this.loading$.next(false);
+                });
+                this.errors$.next([]);
+              },
+              error: error => {
+                this.change$.pipe(take(1)).subscribe(change => {
+                  this.loading$.next(false);
+                });
+                this.handleFormError(error);
+              },
+            });
+          });
+      }
+    });
+  }
+
+  private handleChanges() {
+    this.change$
+      .pipe(
+        pairwise(),
+        debounceTime(500),
+        filter(
+          ([prevChange, currentChange]) =>
+            prevChange?.changed?.value !== undefined &&
+            currentChange?.changed?.value !== undefined &&
+            prevChange.changed.value !== currentChange.changed.value
+        )
+      )
```

Comment:
Can't this be a distinctUntilChanged?

Reply:
It can but then the call gets repeated a bunch and I don't want that